lib/ruby-push-notifications/mpns/mpns_connection.rb

```diff
@@ -0,0 +1,71 @@
+require 'uri'
+require 'net/https'
+
+module RubyPushNotifications
+  module MPNS
+    # Encapsulates a connection to the MPNS service
+    # Responsible for final connection with the service.
+    #
+    class MPNSConnection
+
+      # @private Content-Type HTTP Header string
+      CONTENT_TYPE_HEADER  = 'Content-Type'
+
+      # @private text/xml content type
+      X_NOTIFICATION_CLASS = 'X-NotificationClass'
+
+      # @private Windows Phone Target
+      X_WINDOWSPHONE_TARGET = 'X-WindowsPhone-Target'
+
+      # @private text/xml content type
+      XML_CONTENT_TYPE     = 'text/xml'
+
+      # @private Enumators for notification types
+      BASEBATCH = { tile: 1, toast: 2, raw: 3 }
+
+      # @private Enumators for delay
+      BATCHADDS = { delay450: 10, delay900: 20 }
+
+      # @private Windows Phone Target Types
+      WP_TARGETS = { toast: 'toast', tile: 'token' }
+
+      # Issues a POST request to the MPNS send endpoint to
+      # submit the given notifications.
+      #
+      # @param n [MPNSNotification]. The notification object to POST
+      # @param optional cert [String]. Contents of the PEM encoded certificate
+      # @return [Array]. The response of post
+      # (http://msdn.microsoft.com/pt-br/library/windows/apps/ff941099)
+      def self.post(n, cert = nil)
+        headers = build_headers(n.data[:type], n.data[:delay])
+        body = n.as_mpns_xml
+        responses = []
+        n.each_device do |url|
+          http = Net::HTTP.new url.host, url.port
+          if cert && url.scheme == 'https'
+            http.use_ssl = true
+            http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+            http.ca_file = cert
+          end
+          response = http.post url.path, body, headers
+          responses << { device_url: url.to_s, headers: response, code: response.code.to_i }
```

Comment:
Either return the whole response (removing the code as is redundant), or extract the headers from the response and that way the response object can be freed (preferred option)